dev-docs/API_VERSIONING.md

```diff
@@ -0,0 +1,45 @@
+## API v0 and v1 functionality for users
+
+### v0
+
+#### Expectation
+
+- The user can edit all information, including their public\_key via v0 api (which knife is currently using).
+- If the user performs an update that includes a nil public key, delete the default key from the keys table for that user.
+
+#### Backend
+
+- Since both the users and the keys table had the record of the public\_key there were some issues when the keys are deleted and recreated. Triggers updated the information into the keys table from the users table. (Could not be done with fk because both the user and client table reference the keys table).
+
+#### Bug
+
+- Since every v0 API command overwrites the public\_key in the users table, triggers, and copies the incorrect key onto the correct key in the keys table. This bug is fixed with `keys table is the only source if truth for public_key`.
+
+#### Current Status
+
+- Currently, phase1 is in rollout.
+- knife uses v0.
+
+### v1
+
+- The user is not allowed to edit the public\_key via the /users endpoint, it must be managed via the /keys endpoint instead.
+- chef-server-ctl uses v1.
+
+## Keys table is the only source of truth for public\_key
+
+### Phase1(Currently implemented)
+
+- Add a function for add\_user and update\_user that inserts only the sentinel value into the public\_key field in the users table.
+- The add and update triggers in keys\_update\_trigger.sql are present to maintain compatibility with older versions of chef-server.
+
+### Phase2(To be implemented)
+
+#### Todo
+
+- Delete and update triggers in keys\_update\_trigger.sql.
+- Delete the public\_key column in the users (and perhaps clients) table.
+
+Note:
+- This should be easier once all the users are on Phase1 release of chef-server.
```

Comment:
do we know which is the phase 1 release?

Reply:
The trail ran cold with https://github.com/chef/chef-server/pull/1383